Source/Core/Core/IOS/DI/DI.cpp

```diff
@@ -30,93 +52,658 @@ void DI::DoState(PointerWrap& p)
-    const DiscIO::Partition partition(partition_offset);
-    DVDInterface::ChangePartition(partition);
+    ChangePartition(DiscIO::Partition(partition_offset));
 
-    INFO_LOG(IOS_DI, "DVDLowOpenPartition: partition_offset 0x%016" PRIx64, partition_offset);
+    INFO_LOG(IOS_DI, "DVDLowOpenPartition: partition_offset 0x%09" PRIx64, partition_offset);
 
     // Read TMD to the buffer
-    const IOS::ES::TMDReader tmd = DVDThread::GetTMD(partition);
+    const IOS::ES::TMDReader tmd = DVDThread::GetTMD(m_current_partition);
     const std::vector<u8>& raw_tmd = tmd.GetBytes();
     Memory::CopyToEmu(request.io_vectors[0].address, raw_tmd.data(), raw_tmd.size());
-    m_ios.GetES()->DIVerify(tmd, DVDThread::GetTicket(partition));
 
-    return_value = 1;
+    ReturnCode es_result = m_ios.GetES()->DIVerify(tmd, DVDThread::GetTicket(m_current_partition));
+    Memory::Write_U32(es_result, request.io_vectors[1].address);
+
+    return_value = DIResult::SUCCESS;
     break;
   }
+  case DIIoctl::DVDLowGetNoDiscOpenPartitionParams:
+    ERROR_LOG(IOS_DI, "DVDLowGetNoDiscOpenPartitionParams - dummied out");
+    DolphinAnalytics::Instance().ReportGameQuirk(GameQuirk::USES_DIFFERENT_PARTITION_COMMAND);
+    request.DumpUnknown(GetDeviceName(), LogTypes::IOS_DI);
+    break;
+  case DIIoctl::DVDLowNoDiscOpenPartition:
+    ERROR_LOG(IOS_DI, "DVDLowNoDiscOpenPartition - dummied out");
+    DolphinAnalytics::Instance().ReportGameQuirk(GameQuirk::USES_DIFFERENT_PARTITION_COMMAND);
+    request.DumpUnknown(GetDeviceName(), LogTypes::IOS_DI);
+    break;
+  case DIIoctl::DVDLowGetNoDiscBufferSizes:
+    ERROR_LOG(IOS_DI, "DVDLowGetNoDiscBufferSizes - dummied out");
+    DolphinAnalytics::Instance().ReportGameQuirk(GameQuirk::USES_DIFFERENT_PARTITION_COMMAND);
+    request.DumpUnknown(GetDeviceName(), LogTypes::IOS_DI);
+    break;
+  case DIIoctl::DVDLowOpenPartitionWithTmdAndTicket:
+    ERROR_LOG(IOS_DI, "DVDLowOpenPartitionWithTmdAndTicket - not implemented");
+    DolphinAnalytics::Instance().ReportGameQuirk(GameQuirk::USES_DIFFERENT_PARTITION_COMMAND);
+    break;
+  case DIIoctl::DVDLowOpenPartitionWithTmdAndTicketView:
+    ERROR_LOG(IOS_DI, "DVDLowOpenPartitionWithTmdAndTicketView - not implemented");
+    DolphinAnalytics::Instance().ReportGameQuirk(GameQuirk::USES_DIFFERENT_PARTITION_COMMAND);
+    break;
   default:
+    ERROR_LOG(IOS_DI, "Unknown ioctlv 0x%02x", request.request);
     request.DumpUnknown(GetDeviceName(), LogTypes::IOS_DI);
   }
-  return GetDefaultReply(return_value);
+  return GetDefaultReply(static_cast<s32>(return_value));
+}
+
+void DI::ChangePartition(const DiscIO::Partition partition)
+{
+  m_current_partition = partition;
+}
+
+DiscIO::Partition DI::GetCurrentPartition()
+{
+  auto di = IOS::HLE::GetIOS()->GetDeviceByName("/dev/di");
+  if (di)
+    return std::static_pointer_cast<DI>(di)->m_current_partition;
+  else
+    return DiscIO::PARTITION_NONE;
+}
+
+void DI::InitializeIfFirstTime()
```

Comment:
Consider adding a comment to explain that this is the equivalent of Nintendo DI's initDvdDriverStage2, which is called the first time an open/ioctl/ioctlv IPC command is received. (Otherwise, people might be tempted to move this one-time initialisation code into the constructor, which would be inaccurate)

Reply:
Done.  One thing that's been bothering me, though, is the ioctl/ioctlv cases; can those actually be hit (on console)?  I feel like the device would need to be opened first before an ioctl could be issued, so there's no reason other than paranoia to also check in ioctl/ioctlv.